### PR TITLE
Introduce an end-to-end CI job on Linux

### PR DESCRIPTION
**Pull request by stefwalter**
This PR introduces a new job that runs in CI that performs a minimal
configuration of the `ilab` workflow:

- `ilab init`
- `ilab download`
- `ilab serve`
- `ilab generate`
- `ilab train`

It runs on a GPU-enabled github actions runner. This is a Linux VM
with a single Tesla T4 GPU.

Given the resources required to run this job, we do not propose that
it runs automatically on every PR. Instead, it is a workflow that must
be manually launched. When launching it manually, you can specify
which branch or pull request to run it against.

Signed-off-by: Stef Walter <stefw@redhat.com>
Co-authored-by: Russell Bryant <rbryant@redhat.com>
Signed-off-by: Russell Bryant <rbryant@redhat.com>

---

Depends on:

- [x] #1050 
- [x] #1058 

Related:

- [ ] Scope of what can be tested is limited by #579 

---

Follow-up work after this PR:

- Make another variant of this job that builds our instructlab container with cuda support and uses that to run the same workflow
- Consider how and where to run end-to-end tests for other environments, file tracking issues for those

### CODE REVIEW

**Comment by stefwalter** [comment]
@rhatdan @jeremyeder @lmilbaum FYI

**Comment by stefwalter** [comment]
@lmilbaum @rhatdan These is an e2e test we'd run in https://github.com/containers/ai-lab-recipes/ ... related to https://github.com/containers/ai-lab-recipes/pull/341

**Comment by stefwalter** [comment]
This work needs https://github.com/instructlab/instructlab/pull/1018 to fix the container first.

The work is incomplete right now, with lots of TODO statements indicating what remains to be done.

**Comment by stefwalter** [comment]
@russellb The Container file also needs this for cuda:

```
-RUN dnf install -y libcudnn8 nvidia-driver-NVML
+RUN dnf install -y libcudnn8 nvidia-driver-NVML nvidia-driver-cuda-libs
```

**Comment by stefwalter** [comment]
Is this ready for broader review and testing? Did a CI run work successfully in GitHub?

**Comment by russellb** [comment]
> Is this ready for broader review and testing? Did a CI run work successfully in GitHub?

I don’t have the CI job passing yet. I’ll update again later today.

**Comment by russellb** [comment]
My last push included a rebase to be on top of a bunch of newer changes.  Some outstanding issues:

* I defined an `e2e` CI job, but it doesn't work reliably. I don't think the github action runners have enough CPU to finish even the most minimal configuration. I'm going to switch to looking at running it on a GPU runner instead. The tradeoff will be that it won't run on every PR for now, since it's not free.
* Eventually, we'll squash all of this, but that's for when it's ready for review and merge ... I'd like to have a working CI job ready for this repo first though

**Comment by russellb** [comment]
The end-to-end CI job that does init/download/generate/train on Linux is passing. This is running on the host OS using a GPU worker available via github actions.

Next steps:

- I'm going to clean up this branch by squashing most of it together and putting it in a form ready for review and merging this first workflow
- Next I want to add a variant that instead of running it on the host, builds a cuda container using our containerfile and runs the same workflow using the container. I'd like to do that in a follow-up PR, though.

**Comment by russellb** [comment]
I applied the `hold` label because:

- This PR includes #1058, so that should be reviewed first
- There is a commit on the branch that must be removed before merging (the change to make the job run automatically against this PR). The intent is to only merge it as a manually triggered workflow.

File: src/instructlab/train/linux_train.py
Comment (russellb) [inline src/instructlab/train/linux_train.py:32]: @tiran Do you have any advice on this one? I hit this reliably in this CI job.

Comment (russellb) [reply src/instructlab/train/linux_train.py:32]: The commit message here has the backtrace: https://github.com/instructlab/instructlab/pull/1016/commits/2865268a3c51e8705430164ff87741c24f5f8589

**Comment by russellb** [comment]
Note that the workflow is currently limited and stops before ensuring you can run `ilab serve` with a model after `ilab train` because of #579

**Comment by russellb** [comment]
A data point on costs. The last two successful runs took 22 minutes. That costs us $1.54 each time ($0.07 per minute). If we run this only manually, I think the costs will stay under control.

I also think we could speed this up. A lot of time is spent on downloading and installing stuff. No caching is done in the workflow, yet.

File: src/instructlab/train/linux_train.py
Comment (tiran) [reply src/instructlab/train/linux_train.py:32]: I have seen the issue before on HPUs and created #1050

**Comment by tiran** [comment]
For caching I recommend to use `setup-python` with cache pip and `actions/cache` to cache the Hugging Face models. We are using the same technique in `test.yaml`. You could also cache the APT packages with something like https://github.com/marketplace/actions/cache-apt-packages

```yaml
      - name: Setup Python 3.11
        uses: actions/setup-python@v5
        with:
          python-version: py311
          cache: pip
          cache-dependency-path: |
            **/pyproject.toml
            **/requirements*.txt

      - name: Cache huggingface
        uses: actions/cache@v4
        with:
          path: ~/.cache/huggingface
          # config contains DEFAULT_MODEL
          key: huggingface-${{ hashFiles('src/instructlab/config.py') }}
```

**Comment by tiran** [comment]
Are you sure the PR uses a CUDA-accelerated build of llama-cpp-python? It downloads and builds llama-cpp-python two times:

```
Collecting llama-cpp-python
  Downloading llama_cpp_python-0.2.68.tar.gz (42.5 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 42.5/42.5 MB 170.8 MB/s eta 0:00:00
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
  Installing backend dependencies: started
  Installing backend dependencies: finished with status 'done'
  Preparing metadata (pyproject.toml): started
  Preparing metadata (pyproject.toml): finished with status 'done'
```

```
 Collecting llama_cpp_python[server]==0.2.55
  Downloading llama_cpp_python-0.2.55.tar.gz (36.8 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 36.8/36.8 MB 172.1 MB/s eta 0:00:00
  Installing build dependencies: started
  Installing build dependencies: finished with status 'done'
  Getting requirements to build wheel: started
  Getting requirements to build wheel: finished with status 'done'
  Installing backend dependencies: started
  Installing backend dependencies: finished with status 'done'
  Preparing metadata (pyproject.toml): started
  Preparing metadata (pyproject.toml): finished with status 'done'
```

I think you are overriding the CUDA build with a CPU-only build. Our installation instructions have sharp edges.

Please try this:

```shell
sed 's/\[.*\]//' requirements.txt constraints.txt
python3 -m pip cache remove llama_cpp_python
python3 -m pip install --no-binary llama_cpp_python -c constraints.txt llama_cpp_python
```

File: src/instructlab/train/linux_train.py
Comment (russellb) [reply src/instructlab/train/linux_train.py]: Thanks! I replaced my change with the commit from your PR. Assuming the job still works, I'll approve that one.

**Comment by russellb** [comment]
> Are you sure the PR uses a CUDA-accelerated build of llama-cpp-python?

I guess I'm not! I was going off of the output from `ilab train` confirming that it was using the GPU, but I wasn't thinking that `ilab serve` might be different. Thank you.

File: src/instructlab/train/linux_train.py
Comment (russellb) [reply src/instructlab/train/linux_train.py]: It doesn't fix it here.

```
Traceback (most recent call last):
  File "/home/runner/work/instructlab/instructlab/venv/lib/python3.10/site-packages/instructlab/train/linux_train.py", line 30, in <module>
    torch.multiprocessing.set_start_method(DEFAULT_MULTIPROCESSING_START_METHOD)
  File "/usr/lib/python3.10/multiprocessing/context.py", line 247, in set_start_method
    raise RuntimeError('context has already been set')
RuntimeError: context has already been set

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/runner/work/instructlab/instructlab/venv/bin/ilab", line 8, in <module>
    sys.exit(cli())
  File "/home/runner/work/instructlab/instructlab/venv/lib/python3.10/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
  File "/home/runner/work/instructlab/instructlab/venv/lib/python3.10/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
  File "/home/runner/work/instructlab/instructlab/venv/lib/python3.10/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/runner/work/instructlab/instructlab/venv/lib/python3.10/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/runner/work/instructlab/instructlab/venv/lib/python3.10/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
  File "/home/runner/work/instructlab/instructlab/venv/lib/python3.10/site-packages/click/decorators.py", line 33, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/home/runner/work/instructlab/instructlab/venv/lib/python3.10/site-packages/instructlab/lab.py", line 951, in train
    from .train.linux_train import linux_train
  File "/home/runner/work/instructlab/instructlab/venv/lib/python3.10/site-packages/instructlab/train/linux_train.py", line 34, in <module>
    raise ValueError(
ValueError: multiprocessing start method already set to fork.

```

**Review by nathan-weinberg — APPROVED** [review]
LGTM - I'd love if we could incorporate this into the release workflow somehow - maybe having an automic trigger when we cut a new release branch for Y-streams or for Backport PRs to an existing release branch (to ensure Z-streams have no regressions)

**Comment by russellb** [comment]
> LGTM - I'd love if we could incorporate this into the release workflow somehow - maybe having an automic trigger when we cut a new release branch for Y-streams or for Backport PRs to an existing release branch (to ensure Z-streams have no regressions)

Yeah, that's a good idea. We could definitely trigger it on tag creation. At the moment, it would be triggering it manually, and using the input field that lets you specify which branch or tag to run against. You't put in `release-v0.15` or `v0.15.0` or whatever you want and it'll run for you.

**Pull request merged**
